SourceBuffer.timestampOffset usage doesn't behave correctly with webm content.

jyavenard opened the pull request:
#### bee529142266d4fd8fcc2f587076fba341ad0684
<pre>
SourceBuffer.timestampOffset usage doesn&apos;t behave correctly with webm content.
<a href="https://bugs.webkit.org/show_bug.cgi?id=252782">https://bugs.webkit.org/show_bug.cgi?id=252782</a>
rdar://105801920

Reviewed by Youenn Fablet.

The libwebmparser enforced checking the size of the WebM Segment parent.
When we append a cluster a second time, it would have made the total
size of the webm larger than the parent Segment could fit and the parser
aborted with an error. That error wasn&apos;t propagated to JS.

Considering a webm is only ever made of one Segment, and that the Matroska
specification allows for an element size to be unknown and other browsers
ignore that size, the simplest way to get around is to ignore the
Segment&apos;s size and assume it is always unknown.

This allows to add as many cluster as one wants, regardless of the init
segment added.
libwebm is a 3rd party library, however we have already modified it to
suits our requirement.

* LayoutTests/media/media-source/content/test-opus-manifest.json:
* LayoutTests/media/media-source/content/test-opus.webm: Remux the file with ffmpeg so it reproduces the same issue as the vorbis one.
* LayoutTests/media/media-source/media-source-webm-fragment-expected.txt: Added.
* LayoutTests/media/media-source/media-source-webm-fragment.html: Added.
* Source/ThirdParty/libwebrtc/Source/third_party/libwebm/webm_parser/src/webm_parser.cc:

Canonical link: <a href="https://commits.webkit.org/260822@main">https://commits.webkit.org/260822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd8d8e543bbea55f2c4bf1a273dc450cc7a1d94b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109561 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18684 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1050 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9881 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101830 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115317 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98227 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43209 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29880 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84979 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11420 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31223 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12079 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17444 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50829 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7499 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13818 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/ThirdParty/libwebrtc/Source/third_party/libwebm/webm_parser/src/webm_parser.cc:
```diff
@@ -75,7 +75,7 @@ class WebmParser::DocumentParser {
           }
           child_metadata_.id = id_parser_.id();
-          child_metadata_.size = size_parser_.size();
+          child_metadata_.size = child_metadata_.id == Id::kSegment ? kUnknownElementSize : size_parser_.size();
```
We probably need to keep track of changes like this since we might not upstream the changes.
In libwebrtc, all webkit specific changes are marked with `#if defined(WEBRTC_WEBKIT_BUILD)`

jyavenard replied:
new test would fail if it happened. I will add a similar macro to flag a webkit only change

jyavenard replied:
done

jyavenard committed:
SourceBuffer.timestampOffset usage doesn't behave correctly with webm content.
https://bugs.webkit.org/show_bug.cgi?id=252782
rdar://105801920

Reviewed by Youenn Fablet.

The libwebmparser enforced checking the size of the WebM Segment parent.
When we append a cluster a second time, it would have made the total
size of the webm larger than the parent Segment could fit and the parser
aborted with an error. That error wasn't propagated to JS.

Considering a webm is only ever made of one Segment, and that the Matroska
specification allows for an element size to be unknown and other browsers
ignore that size, the simplest way to get around is to ignore the
Segment's size and assume it is always unknown.

This allows to add as many cluster as one wants, regardless of the init
segment added.
libwebm is a 3rd party library, however we have already modified it to
suits our requirement.

* LayoutTests/media/media-source/content/test-opus-manifest.json:
* LayoutTests/media/media-source/content/test-opus.webm: Remux the file with ffmpeg so it reproduces the same issue as the vorbis one.
* LayoutTests/media/media-source/media-source-webm-fragment-expected.txt: Added.
* LayoutTests/media/media-source/media-source-webm-fragment.html: Added.
* Source/ThirdParty/libwebrtc/Source/third_party/libwebm/webm_parser/src/webm_parser.cc:

Canonical link: https://commits.webkit.org/260822@main
